pkg/cache/tas_flavor_snapshot.go

```diff
@@ -0,0 +1,349 @@
+	s.initializeFreeCapacityPerDomain(domainID)
+	s.freeCapacityPerDomain[domainID].Sub(usage)
+}
+
+func (s *TASFlavorSnapshot) initializeFreeCapacityPerDomain(domainID utiltas.TopologyDomainID) {
+	if _, found := s.freeCapacityPerDomain[domainID]; !found {
+		s.freeCapacityPerDomain[domainID] = resources.Requests{}
+	}
+}
+
+// Algorithm overview:
+// Phase 1:
+//
+//	determine pod counts for each topology domain. Start at the lowest level
+//	and bubble up the numbers to the top level
+//
+// Phase 2:
+//
+//	a) select the domain at requested level with count >= requestedCount
+//	b) traverse the structure down level-by-level optimizing the number of used
+//	  domains at each level
+func (s *TASFlavorSnapshot) FindTopologyAssignment(
+	topologyRequest *kueue.PodSetTopologyRequest,
+	requests resources.Requests,
+	count int32) *kueue.TopologyAssignment {
+	required := topologyRequest.Required != nil
+	levelIdx, found := s.resolveLevelIdx(topologyRequest)
+	if !found {
+		return nil
+	}
+	// phase 1 - determine the number of pods which can fit in each topology domain
+	s.fillInCounts(requests)
+
+	// phase 2a: determine the level at which the assignment is done along with
+	// the domains which can accommodate all pods
+	fitLevelIdx, currFitDomain := s.findLevelWithFitDomains(levelIdx, required, count)
+	if len(currFitDomain) == 0 {
+		return nil
+	}
+
+	// phase 2b: traverse the tree down level-by-level optimizing the number of
+	// topology domains at each level
+	currFitDomain = s.updateCountsToMinimum(currFitDomain, count)
+	for levelIdx := fitLevelIdx; levelIdx+1 < len(s.domainsPerLevel); levelIdx++ {
+		lowerFitDomains := s.lowerLevelDomains(levelIdx, currFitDomain)
+		sortedLowerDomains := s.sortedDomains(lowerFitDomains)
+		currFitDomain = s.updateCountsToMinimum(sortedLowerDomains, count)
+	}
+	return s.buildAssignment(currFitDomain)
+}
+
+func (s *TASFlavorSnapshot) resolveLevelIdx(
+	topologyRequest *kueue.PodSetTopologyRequest) (int, bool) {
+	var levelKey string
+	if topologyRequest.Required != nil {
+		levelKey = *topologyRequest.Required
+	} else if topologyRequest.Preferred != nil {
+		levelKey = *topologyRequest.Preferred
+	}
+	levelIdx := slices.Index(s.levelKeys, levelKey)
+	if levelIdx == -1 {
+		return levelIdx, false
+	}
+	return levelIdx, true
+}
+
+func (s *TASFlavorSnapshot) findLevelWithFitDomains(levelIdx int, required bool, count int32) (int, []*domain) {
+	levelDomains := s.domainsForLevel(levelIdx)
+	if len(levelDomains) == 0 {
+		return 0, nil
+	}
+	sortedDomain := s.sortedDomains(levelDomains)
+	topDomain := sortedDomain[0]
+	if s.state[topDomain.id] < count {
+		if required {
+			return 0, nil
+		}
+		if levelIdx > 0 {
+			return s.findLevelWithFitDomains(levelIdx-1, required, count)
+		}
+		lastIdx := 0
+		remainingCount := count - s.state[sortedDomain[lastIdx].id]
+		for remainingCount > 0 && lastIdx < len(sortedDomain)-1 {
```

Comment:
Can we check if the `s.state[sortedDomain[lastIdx].id] > 0` to reduce adding domains that can't fit any pod?

Reply:
Done in the follow up https://github.com/kubernetes-sigs/kueue/pull/3291